File: JAVA_Programming/SinglyLL.java
// Singly Linear Linked List

class node
{
    public int Data;
    public node Next;

    public node(int no)
    {
        this.Data = no;
        this.Next = null;
    }
}

class SinglyLinear
{
    public node first;
    public int iCount;

    public SinglyLinear()
    {
        System.out.println("Object of SinglyLinear Created.");
        this.first = null;
        this.iCount = 0;
    }

    public void InsertFirst(int no)
    {
        node newn = null;
        newn = new node(no);

        newn.Next = this.first;
        this.first = newn;
        
        this.iCount++;
    }

    public void InsertLast(int no)
    {
        node temp = null;
        node newn = null;
        newn = new node(no);
 
        if(this.first == null)
        {
            this.first = newn;
        }
        else
        {
            temp = this.first;
            while(temp.Next != null)
            {
                temp = temp.Next;
            }
            temp.Next = newn;
        }
        this.iCount++;
    }

    public void InsertAtPos(int no, int pos)
    {
        node temp = null;
        node newn = null;
        newn = new node(no);

        if(pos < 1 || pos > iCount +1)
        {
            System.out.println("Invalid Position.");
            return;
        }

        if(pos == 1)
        {
            InsertFirst(no);
        }
        else if(pos == iCount +1)
        {
            InsertLast(no);
        }
        else
        {
            temp = this.first;
            for(int iCnt = 1; iCnt < pos -1; iCnt++)
            {
                temp = temp.Next;
            }
            newn.Next = temp.Next;
            temp.Next = newn;
            this.iCount++;
        }
    }

    public void DeleteFirst()
    {
        if(this.first == null)
        {
            System.out.println("Node is Empty.");
            return;
        }
        else
        {
            this.first = this.first.Next;
        }
        System.gc();
        this.iCount--;       
    }

    public void DeleteLast()
    {
        node temp = null;

        if(this.first == null)
        {
            System.out.println("Node is Empty.");
            return;
        }
        else if(this.first.Next == null)
        {
            this.first = null;
            System.gc();
        }
        else
        {
            temp = this.first;
            while(temp.Next.Next != null)
            {
                temp = temp.Next;
            }
            temp.Next = null;
            System.gc();
        }
        this.iCount--;
    }

    public void DeleteAtPos(int pos)
    {
        node target = null;
        node temp = null;
        if(pos < 1 || pos > iCount)
        {
            System.out.println("Invalid Position.");
            return;
        }

        if(pos == 1)
        {
            DeleteFirst();
        }
        else if(pos == iCount)
        {
            DeleteLast();
        }
        else
        {
            temp = this.first;
            for(int iCnt = 1; iCnt < pos -1; iCnt++)
            {
                temp = temp.Next;
            }
            target = temp.Next;
            temp.Next = target.Next;
            System.gc();            
            this.iCount--;
        }
    }

    public void Display()
    {
        node temp = null;
        temp = this.first;
        while(temp != null)
        {
            System.out.print("| "+temp.Data+" | -> ");
            temp = temp.Next;
        }
        System.out.println("null");        
    }

    public int Count()
    {
        return iCount;
    }
}

class SinglyLL
{
    public static void main(String[] args)
    {
        int iRet = 0;
        SinglyLinear sobj = new SinglyLinear();
    
        System.out.println("----------------------------------------------------------------------------");
        sobj.InsertFirst(90);
        sobj.InsertFirst(60);
        sobj.InsertFirst(30);
        sobj.Display();
        iRet = sobj.Count();
        System.err.println("Number of After Insert First Node are : "+iRet);

        System.out.println("----------------------------------------------------------------------------");
        sobj.InsertLast(33);
        sobj.InsertLast(66);
        sobj.InsertLast(99);
        sobj.Display();
        iRet = sobj.Count();
        System.err.println("Number of After Insert Last Node are : "+iRet);

        System.out.println("----------------------------------------------------------------------------");
        sobj.InsertAtPos(180,4);
        sobj.Display();
        iRet = sobj.Count();
        System.err.println("Number of After Insert At Position Node are : "+iRet);

        System.out.println("----------------------------------------------------------------------------");
        sobj.DeleteFirst();
        sobj.Display();
        iRet = sobj.Count();
        System.err.println("Number of After Delete First Node are : "+iRet);

        System.out.println("----------------------------------------------------------------------------");
        sobj.DeleteLast();
        sobj.Display();
        iRet = sobj.Count();
        System.err.println("Number of After Delete Last Node are : "+iRet);

        System.out.println("----------------------------------------------------------------------------");
        sobj.DeleteAtPos(3);
        sobj.Display();
        iRet = sobj.Count();
        System.err.println("Number of After Delete At Position Node are : "+iRet);
    }
}

File: C++_Programming/SinglyLL.cpp
// Singly Linear Linked List

#include<iostream>
using namespace std;

struct node
{
    int Data;
    struct node *Next;
};

typedef struct node NODE;
typedef struct node* PNODE;


class SinglyLL
{
    private :
        PNODE first;
        int iCount;

    public :
        SinglyLL()
        {
            cout<<"Object of SinglyLL is Created..\n";
            this->first = NULL;
            this->iCount = 0;
        }
        void InsertFirst(int no)
        {
            PNODE newn = NULL;
            newn = new NODE;

            newn->Data = no;
            newn->Next = NULL;

            newn->Next = this->first;
            this->first = newn;
            this->iCount++; 
        }

        void InsertLast(int no)
        {
            PNODE newn = NULL;
            PNODE temp = NULL;

            newn = new NODE;

            newn->Data = no;
            newn->Next = NULL;

            if(this->first == NULL)
            {
                this->first = newn;
            }
            else
            {
                temp = this -> first;
                while(temp -> Next != NULL)
                {
                    temp = temp -> Next;
                }
                temp->Next = newn;
            }
            this->iCount++;
        }

        void InsertAtPos(int no, int pos)
        {
            PNODE newn = NULL;
            PNODE temp = NULL;
            
            if(pos < 1 || pos > iCount +1)
            {
                cout<<"Invalid Position...\n";
                return;
            }

            if(pos == 1)
            {
                InsertFirst(no);
            }
            else if(pos == iCount +1)
            {
                InsertLast(no);
            }
            else
            {
                newn = new NODE;

                newn->Data = no;
                newn->Next = NULL;

                temp = first;
                for(int iCnt = 1; iCnt < pos -1; iCnt++)
                {
                    temp = temp -> Next;
                }
                newn->Next = temp ->Next;
                temp->Next = newn;
                this->iCount++;
            }
        }

        void DeleteFirst()
        {
            PNODE temp = NULL;
            if(this->first == NULL)
            {
                return;
            }
            else if(this->first->Next == NULL)
            {
                delete this -> first;
                this->first = NULL;
            }
            else
            {
                temp = this -> first;                

                this->first = this->first->Next;
                delete temp;
            }
            this -> iCount--;
        }

        void DeleteLast()
        {
            PNODE temp = NULL;
            if(this->first == NULL)
            {
                return;
            }
            else if(this->first->Next == NULL)
            {
                delete this->first;
                this->first = NULL;
            }
            else
            {
                temp = this->first;
                while (temp->Next->Next != NULL)
                {
                    temp = temp->Next;
                }
                delete temp->Next;
                temp->Next = NULL;
                this->iCount--;
            }
        }

        void DeleteAtPos(int pos)
        {
            PNODE temp = NULL;
            PNODE target = NULL;

            if(pos < 1 || pos > iCount)
            {
                cout<<"Invalid Position...\n";
                return;
            }
            
            if(pos == 1)
            {
                DeleteFirst();
            }
            else if(pos == iCount)
            {
                DeleteLast();
            }
            else
            {
                temp = this->first;
                for(int iCnt = 1; iCnt < pos -1; iCnt++)
                {
                    temp = temp->Next;
                }
                target = temp->Next;
                temp->Next = target;
                delete target;
                this->iCount--;
            }
        }

        void Display()
        {
            PNODE temp = NULL;
            temp = this->first;
            while(temp != NULL)
            {
                cout<<"| "<<temp->Data<<" | -> ";
                temp = temp -> Next;
            }
            cout<<"NULL\n";
        }
        int Count()
        {
            return iCount;
        }
};

int main()
{
    int iRet = 0;
    SinglyLL obj;

    obj.InsertFirst(27);
    obj.InsertFirst(24);
    obj.InsertFirst(21);
    obj.Display();
    iRet = obj.Count();
    cout<<"Numbers of After Insert First Node are : "<<iRet<<"\n";
    cout<<"-----------------------------------------------------------\n";

    obj.InsertLast(30);
    obj.InsertLast(33);
    obj.InsertLast(39);
    obj.Display();
    iRet = obj.Count();
    cout<<"Numbers of After Insert Last Node are : "<<iRet<<"\n";
    cout<<"-----------------------------------------------------------\n";
    
    obj.InsertAtPos(36,6);
    obj.Display();
    iRet = obj.Count();
    cout<<"Numbers of After InsertAt Position Node are : "<<iRet<<"\n";
    cout<<"-----------------------------------------------------------\n";
    
    obj.DeleteFirst();
    obj.Display();
    iRet = obj.Count();
    cout<<"Numbers of After Delete First Node are : "<<iRet<<"\n";
    cout<<"-----------------------------------------------------------\n";

    obj.DeleteLast();
    obj.Display();
    iRet = obj.Count();
    cout<<"Numbers of After Delete Last Node are : "<<iRet<<"\n";
    cout<<"-----------------------------------------------------------\n";
    
    obj.DeleteAtPos(6);
    obj.Display();
    iRet = obj.Count();
    cout<<"Numbers of After Delete At Position Node are : "<<iRet<<"\n";
    cout<<"-----------------------------------------------------------\n";
    
    return 0;
}
